File: scripts/solar/util/game/CameraUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["6"] = 1,["7"] = 1,["8"] = 1,["10"] = 1,["11"] = 19,["12"] = 19,["13"] = 19,["15"] = 20,["16"] = 21,["17"] = 22,["19"] = 25,["20"] = 26,["21"] = 27,["22"] = 28,["23"] = 29,["24"] = 30,["25"] = 31,["26"] = 33,["27"] = 35,["28"] = 36,["29"] = 37,["30"] = 39,["31"] = 40,["32"] = 41,["33"] = 42,["34"] = 43,["35"] = 19,["36"] = 74,["37"] = 75,["38"] = 76,["39"] = 77,["42"] = 80,["44"] = 74,["45"] = 88,["46"] = 88,["47"] = 88,["49"] = 89,["50"] = 90,["51"] = 91,["54"] = 94,["56"] = 88,["57"] = 104,["58"] = 104,["59"] = 104,["61"] = 104,["62"] = 104,["64"] = 105,["65"] = 106,["66"] = 106,["67"] = 106,["68"] = 106,["69"] = 106,["71"] = 104,["72"] = 1});
local ____exports = {}
____exports.default = __TS__Class()
local CameraUtil = ____exports.default
CameraUtil.name = "CameraUtil"
function CameraUtil.prototype.____constructor(self)
end
function CameraUtil.getScreenCoordinates(self, x, y, z)
    if z == nil then
        z = 150
    end
    if isEmbedJapi then
        local sx, sy, ss = world_to_screen(x, y, z)
        return {x = sx, y = 0.6 - sy}
    end
    local y1 = GetCameraEyePositionY()
    local x1 = GetCameraEyePositionX()
    local z1 = GetCameraEyePositionZ()
    local w = (DzGetWindowWidth() - 20) / 2
    local h = DzGetWindowHeight() - 20
    local a = GetCameraField(ConvertCameraField(3)) * 180 / 3.141592653589793
    local b = 360 - GetCameraField(ConvertCameraField(2)) * 180 / 3.141592653589793
    y = y + z * (y - y1) / z1
    local sy = 0.3 * (y - y1 - z1 / math.tan(b + a / 2)) * math.sin(b + a / 2) / (z1 * math.sin(a / 2) * (math.sin(b) + math.cos(b) * (y - y1) / z1))
    sy = (sy - 0.3) * 1.66 + 0.3
    local sx = 0.4 * (x - x1) / w / (1 + 1 / ((y - y1) * (z1 / (y - y1) * math.tan(b) + 1) / (y - y1 - z1 / math.tan(b + a / 2)) - 1))
    local c = 235
    local d = 56 * c - 10500
    local sx1 = 0.03 * (w + b * c - d) / h
    sx = (sx + sx1) * 0.4 / sx1
    return {x = sx, y = sy}
end
function CameraUtil.setCameraHeight(self, height, whichPlayer)
    if whichPlayer then
        if GetLocalPlayer() == whichPlayer then
            SetCameraField(CAMERA_FIELD_TARGET_DISTANCE, height, 0)
        end
    else
        SetCameraField(CAMERA_FIELD_TARGET_DISTANCE, height, 0)
    end
end
function CameraUtil.moveCamera2xy(self, x, y, whichPlayer, duration)
    if duration == nil then
        duration = 0
    end
    if whichPlayer then
        if GetLocalPlayer() == whichPlayer then
            PanCameraToTimed(x, y, duration)
        end
    else
        PanCameraToTimed(x, y, duration)
    end
end
function CameraUtil.moveCamera2unit(self, unit, whichPlayer, duration)
    if whichPlayer == nil then
        whichPlayer = GetOwningPlayer(unit)
    end
    if duration == nil then
        duration = 0
    end
    if GetLocalPlayer() == whichPlayer then
        PanCameraToTimed(
            GetUnitX(unit),
            GetUnitY(unit),
            duration
        )
    end
end
____exports.default = CameraUtil
return ____exports

File: src/solar/util/game/CameraUtil.ts
export default class CameraUtil {


    /*
配合单位血条高度 可以模拟血条
配合UI文本 可以实现自定义的字体的UI跳字
配合UI 可以实现在多个掉落物品上提示文本 传奇那种装备爆满地
配合UI 可以画英雄称号 商店标题 等等
     */
    /**
     * 根据世界坐标转屏幕坐标
     * @param x
     * @param y
     * @param z
     * 返回值
     * 屏幕x轴 左到右 0 ~ 0.8
     * 屏幕y轴 下到上 0 ~ 0.6
     */
    static getScreenCoordinates(x: number, y: number, z: number = 150): { x: number, y: number } {
        if (isEmbedJapi) {
            const [sx, sy, ss] = world_to_screen(x, y, z)
            return {x: sx, y: 0.6 - sy}
        }

        let y1 = GetCameraEyePositionY()
        let x1 = GetCameraEyePositionX()
        let z1 = GetCameraEyePositionZ()
        let w = (DzGetWindowWidth() - 20) / 2
        let h = (DzGetWindowHeight() - 20)
        let a = GetCameraField(ConvertCameraField(3)) * 180 / 3.14159265358979323846
        let b = 360 - GetCameraField(ConvertCameraField(2)) * 180 / 3.14159265358979323846

        y += z * (y - y1) / z1

        let sy = 0.3 * (y - y1 - (z1 / Math.tan(b + a / 2))) * Math.sin(b + a / 2) / (z1 * Math.sin(a / 2) * (Math.sin(b) + (Math.cos(b) * (y - y1) / z1)))
        sy = (sy - 0.3) * 1.66 + 0.3
        let sx = (0.4 * (x - x1) / w) / (1 + 1 / (((y - y1) * (z1 / (y - y1) * Math.tan(b) + 1) / (y - y1 - (z1 / Math.tan(b + a / 2)))) - 1))

        let c = 235
        let d = 56 * c - 10500
        let sx1 = 0.03 * (w + b * c - d) / h
        sx = (sx + sx1) * 0.4 / sx1
        return {x: sx, y: sy}
    }


    /*
            let a = 16.1//35*0.46
        let b = 56
        let y1 = GetCameraEyePositionY()
        let x1 = GetCameraEyePositionX()
        let z1 = GetCameraEyePositionZ()
        let w = (DzGetWindowWidth() - 20) / 2
        let h = (DzGetWindowHeight() - 20)

        let cy = y + (z * (y - y1) / (z1 - z))
        let cx = x + (z * (x - x1) / (z1 - z))

        let ty = cy - y1

        w = (w * z1) * 0.85 / h

        let sy = 540 * (ty - (z1 / Tan(b + a))) * Sin(b + a) / (z1 * Sin(a) * (Sin(b) + (Cos(b) * (ty) / z1)))
        let sx = (960 * (cx - x1) / w) / (1 + 1 / (((ty) * (z1 / (ty) * Tan(b) + 1) / (ty - (z1 / Tan(b + a)))) - 1)) + 960

        sy = sy * (790 / 1080) + 250

        return {x: sx, y: sy}
     */

    /**
     * 设置镜头高度
     */
    static setCameraHeight(height: number, whichPlayer?: player) {
        if (whichPlayer) {
            if (GetLocalPlayer() == whichPlayer) {
                SetCameraField(CAMERA_FIELD_TARGET_DISTANCE, height, 0)
            }
        } else {
            SetCameraField(CAMERA_FIELD_TARGET_DISTANCE, height, 0)
        }

    }

    /**
     * 平移镜头高度
     */
    static moveCamera2xy(x: number, y: number, whichPlayer?: player, duration = 0) {
        if (whichPlayer) {
            if (GetLocalPlayer() == whichPlayer) {
                PanCameraToTimed(x, y, duration)
            }
        } else {
            PanCameraToTimed(x, y, duration)
        }
    }

    /**
     * 移动今天到单位位置
     * @param unit
     * @param whichPlayer
     * @param duration
     */
    static moveCamera2unit(unit: unit, whichPlayer: player = GetOwningPlayer(unit), duration = 0) {
        if (GetLocalPlayer() == whichPlayer) {
            PanCameraToTimed(GetUnitX(unit), GetUnitY(unit), duration)
        }
    }
}


